### PR TITLE
fix: cabinet comparitor

### PR DESCRIPTION
**Pull request by mfw78**
**Problem**
Previously, the guard for an `epoch` of zero wasn't triggering as the return value of the cabinet was of type `BigNumber` and the `number` comparison to `0` would never trigger due to not being of type `number`.

**Solution**
This PR correctly handles the bignumber to number marshalling and guards against out-of-range conditions. As there is an interface requirement for the cabinet to be set if the order is of start mining time type, any order that doesn't meet this is requirement is automatically removed from the watch-tower.

### CODE REVIEW

File: src/composable/orderTypes/Twap.ts [hunk 1]
```diff
@@ -273,13 +273,21 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
...
+
+    // Convert the cabinet epoch (bignumber) to a number.
+    const cabinetEpoch = rawCabinetEpoch.toNumber()
 
     if (cabinetEpoch === 0) {
       throw new Error('Cabinet is not set. Required for TWAP orders that start at mining time.')
     }
 
-    return parseInt(cabinet, 16)
+    return cabinetEpoch
```
Comment (alfetopito) [inline src/composable/orderTypes/Twap.ts:290]: What about, instead of converting it to a `number` and potentially lose precision, work instead with a BigNumber instance?

Unless, we know this should never exceed default JS number precision, and thus the MAX_UINT32 check above.

**Review by alfetopito — APPROVED** [review]

File: src/composable/orderTypes/Twap.ts (hunk 1 above)
Comment (mfw78) [reply src/composable/orderTypes/Twap.ts:290]: In general, I'd agree with the `BigNumber` approach. The main reason I kept with the `number` is if it's a legitimate TWAP, the TWAP expects a `block.timestamp` to be in the cabinet. As any arbitrary value can _technically_ be in the cabinet (they mess with the `createWithContext` call), it is not _guaranteed_, in which case it will throw and be classified as a malformed / invalid order, otherwise, we know for sure that it will never exceed the JS number precision.

File: src/composable/orderTypes/Twap.ts [hunk 2]
```diff
@@ -273,13 +273,21 @@ export class Twap extends ConditionalOrder<TwapData, TwapStruct> {
     }
 
     const cabinet = await this.cabinet(params)
-    const cabinetEpoch = utils.defaultAbiCoder.decode(['uint256'], cabinet)[0]
+    const rawCabinetEpoch = utils.defaultAbiCoder.decode(['uint256'], cabinet)[0] as BigNumber
+
+    // Guard against out-of-range cabinet epoch
+    if (rawCabinetEpoch.gt(MAX_UINT32)) {
+      throw new Error(`Cabinet epoch out of range: ${rawCabinetEpoch.toString()}`)
+    }
```
Comment (anxolin) [inline src/composable/orderTypes/Twap.ts:281]: Can we add a unit test for a cabinet overflow?

**Review by anxolin — COMMENTED** [review]
The PR looks great, but I think we should include a unit test to repoduce the cabinete overflow case

File: src/composable/orderTypes/Twap.ts (hunk 2 above)
Comment (mfw78) [reply src/composable/orderTypes/Twap.ts:281]: Done.

File: src/composable/orderTypes/Twap.spec.ts [hunk 3]
```diff
@@ -366,6 +365,20 @@ describe('Poll Validate', () => {
     })
   })
 
+  test(`[CABINET OVERFLOW] The cabinet stored value is greater than uint32`, async () => {
+    // GIVEN: The cabinet stored value is greater than uint32
+    mockCabinet.mockReturnValue(uint256Helper(2 ** 32))
```
Comment (anxolin) [inline src/composable/orderTypes/Twap.spec.ts:370]: beautiful

**Review by anxolin — APPROVED** [review]
